Set up a WebRTC App

This should be able to share a PowerPoint. Requires investigation into the [OVE WebRTC App](https://ove.readthedocs.io/en/stable/ove-apps/packages/ove-app-webrtc/README.html)

**Note:**
This issue was originally part of the proof of concept so I will consider this issue complete when a similarly minimally specced WebRTC app can be launched.

Although, I am getting further than before. When I attempt to compose it all together and run the vis system, I can access all the apps except for the WebRTC app (despite not using the most recent `ove-ui` version). The docker logs show this error:

```sh
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.418 pm - webrtc           : Starting application: webrtc
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.450 pm - webrtc           : Application directories: {"base":"/usr/src/app/packages/ove-app-webrtc/dist","nodeModules":"/usr/src/app/node_modules","constants":"/usr/src/app/packages/ove-app-webrtc/dist/client/constants","rootPage":"/usr/src/app/node_modules/@ove-lib/appbase/lib/landing.html"}
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.453 pm - webrtc           : Using CORS middleware
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.461 pm - webrtc           : Using Express JSON middleware
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.475 pm - OVEUtils         : Registering routes for persistence
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.483 pm - OVEUtils         : Building Swagger API documentation
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.560 pm - OVEUtils         : Registering routes for content pages
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.562 pm - OVEUtils         : Got root page: /usr/src/app/node_modules/@ove-lib/appbase/lib/landing.html
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.658 pm - webrtc           : Using module: openvidu-browser
gridlington-vis-ovehub-ove-apps-1                          | [DEBUG] 11/08/2023, 3:31:53.661 pm - webrtc           : Using module: fontawesome-free
gridlington-vis-ovehub-ove-apps-1                          | /usr/src/app/node_modules/openvidu-node-client/lib/OpenVidu.js:674
gridlington-vis-ovehub-ove-apps-1                          |             throw new Error('URL format incorrect: ' + error);
gridlington-vis-ovehub-ove-apps-1                          |             ^
gridlington-vis-ovehub-ove-apps-1                          | Error: URL format incorrect: TypeError [ERR_INVALID_URL]: Invalid URL: 192.168.0.5:4443
gridlington-vis-ovehub-ove-apps-1                          |     at OpenVidu.setHostnameAndPort (/usr/src/app/node_modules/openvidu-node-client/lib/OpenVidu.js:674:19)
gridlington-vis-ovehub-ove-apps-1                          |     at new OpenVidu (/usr/src/app/node_modules/openvidu-node-client/lib/OpenVidu.js:51:14)
gridlington-vis-ovehub-ove-apps-1                          |     at Object.<anonymous> (/usr/src/app/packages/ove-app-webrtc/dist/index.js:33:14)
gridlington-vis-ovehub-ove-apps-1                          |     at Module._compile (internal/modules/cjs/loader.js:1114:14)
gridlington-vis-ovehub-ove-apps-1                          |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1143:10)
gridlington-vis-ovehub-ove-apps-1                          |     at Module.load (internal/modules/cjs/loader.js:979:32)
gridlington-vis-ovehub-ove-apps-1                          |     at Function.Module._load (internal/modules/cjs/loader.js:819:12)
gridlington-vis-ovehub-ove-apps-1                          |     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:75:12)
gridlington-vis-ovehub-ove-apps-1                          |     at internal/main/run_main_module.js:17:47
gridlington-vis-ovehub-ove-apps-1                          | npm
gridlington-vis-ovehub-ove-apps-1                          |  ERR! code ELIFECYCLE
gridlington-vis-ovehub-ove-apps-1                          | npm
gridlington-vis-ovehub-ove-apps-1                          |  ERR! errno 1
gridlington-vis-ovehub-ove-apps-1                          | npm ERR!
gridlington-vis-ovehub-ove-apps-1                          |  @ove/ove-app-webrtc@0.6.4 start: `node dist/index.js --color`
gridlington-vis-ovehub-ove-apps-1                          | npm ERR! Exit status 1
gridlington-vis-ovehub-ove-apps-1                          | npm ERR! 
gridlington-vis-ovehub-ove-apps-1                          | npm ERR! Failed at the @ove/ove-app-webrtc@0.6.4 start script.
gridlington-vis-ovehub-ove-apps-1                          | npm ERR!
gridlington-vis-ovehub-ove-apps-1                          |  This is probably not a problem with npm. There is likely additional logging output above.
gridlington-vis-ovehub-ove-apps-1                          | npm
gridlington-vis-ovehub-ove-apps-1                          |  WARN Local package.json exists, but node_modules missing, did you mean to install?
gridlington-vis-ovehub-ove-apps-1                          | npm ERR! A complete log of this run can be found in:
gridlington-vis-ovehub-ove-apps-1                          | npm ERR!     /root/.npm/_logs/2023-08-11T15_31_53_766Z-debug.log
gridlington-vis-ovehub-ove-apps-1                          | 2023-08-11T15:31:55: PM2 log: App [@ove/ove-app-webrtc:13] exited with code [1] via signal [SIGINT]
gridlington-vis-ovehub-ove-apps-1                          | 2023-08-11T15:31:58: PM2 log: App [@ove/ove-app-webrtc:13] starting in -fork mode-
gridlington-vis-ovehub-ove-apps-1                          | 2023-08-11T15:31:58: PM2 log: App [@ove/ove-app-webrtc:13] online
gridlington-vis-ovehub-ove-apps-1                          | > @ove/ove-app-webrtc@0.6.4 start /usr/src/app/packages/ove-app-webrtc
gridlington-vis-ovehub-ove-apps-1                          | > node dist/index.js --color
```

--- a/README.md
+++ b/README.md
@@ -26,3 +26,25 @@ Utimately, we will want to automate this setup as much as possible so the enviro
 -   server_name "";
 +   server_name liionsden.rcs.ic.ac.uk;
 ```
+
+## Current steps for Running a WebRTC app and sharing an application window
+
+1. Make sure nothings running
+2. Run `run.sh`` for vis system to run config script and docker-compose
+   * For local/develop versions, use locally built images (the run.sh should cover the config for this). Images must be named:
+     * ove-apps:9.9.9
+     * ove-ove:9.9.9
+     * ove-ui:9.9.9
+   * Check get_ip_address is correct
+   * Check ip address from configure.py is running OVE
+3. Go to OVE Core at http://localhost:8080  <!-- markdownlint-disable-line MD034 -->
+4. Select "Launch new OVE Application instances"
+5. Select WebRTC application, fill in settings and launch instance
+6. Start call button on controller and get session ID
+   * Note: closing controller ends the session
+7. Go to openvidu ip address and manually trust website (OPENVIDU_HOST in the docker-compose.yml)
+8. Log in with username: admin, password: (OPENVIDU_SECRET from docker-compose.yml)
+9. Enter session ID (from controller window) as room name and join the call
+10. Share a window and check it can be viewed from other windows
+
+**Note:** Only one screen can be shared from one browser, so multiple screens will require multiple computers/browsers/incognito windows etc.

--- a/config/default.conf
+++ b/config/default.conf
@@ -147,15 +147,6 @@ server {
       proxy_http_version 1.1;
 }
 
-   location /service/persistence/inmemory {
-      rewrite /service/persistence/inmemory/(.*) /$1 break;
-      rewrite /service/persistence/inmemory / break;
-      proxy_pass http://ovehub-ove-service-persistence-inmemory:8190;
-      proxy_set_header Upgrade $http_upgrade;
-      proxy_set_header Connection $connection_upgrade;
-      proxy_http_version 1.1;
-   }
-
    location /ui/launcher {
       rewrite /ui/launcher/(.*) /$1 break;
       rewrite /ui/launcher / break;

--- a/configure.py
+++ b/configure.py
@@ -43,7 +43,6 @@ def generate_docker_compose(template_file: str, ip: str, develop: bool = False) 
     """
     lines_to_replace = {
         "OVE_HOST": f"{ip}:8080",
-        "TOURIS_HOST": f"{ip}:7080",
         "OPENVIDU_HOST": f"{ip}:4443",
         "openvidu.publicurl": f"https://{ip}:4443",
     }
@@ -60,9 +59,6 @@ def generate_docker_compose(template_file: str, ip: str, develop: bool = False) 
     docker_compose["services"]["ovehub-ove-apps"]["environment"][
         "OVE_HOST"
     ] = lines_to_replace["OVE_HOST"]
-    docker_compose["services"]["ovehub-ove-apps"]["environment"][
-        "TOURIS_HOST"
-    ] = lines_to_replace["TOURIS_HOST"]
     docker_compose["services"]["ovehub-ove-apps"]["environment"][
         "OPENVIDU_HOST"
     ] = lines_to_replace["OPENVIDU_HOST"]

--- a/configure.py
+++ b/configure.py
@@ -71,6 +71,9 @@ def generate_docker_compose(template_file: str, ip: str, develop: bool = False) 
     }
     if develop:
         docker_compose["services"]["dash"]["build"] = "."
+        docker_compose["services"]["ovehub-ove-apps"]["image"] = "ove-apps:9.9.9"
+        docker_compose["services"]["ovehub-ove-ove"]["image"] = "ove-ove:9.9.9"
+        docker_compose["services"]["ovehub-ove-ui"]["image"] = "ove-ui:9.9.9"
     else:
         docker_compose["services"]["dash"][
             "image"

--- a/configure.py
+++ b/configure.py
@@ -4,7 +4,7 @@ Run this script to create the docker-compose.yml file using the
 docker-compose.setup.ove.yml file and the IP address of the machine
 Finds the line in docker-compose.setup.ove.yml that contain the host IP
 address and replaces the value with the IP address of the machine in the new
-file. Also replaces the openvidu-call version to 2.12.0.
+file.
 
 Does the same for config/credentials.json file.
 """
@@ -43,8 +43,8 @@ def generate_docker_compose(template_file: str, ip: str, develop: bool = False) 
     """
     lines_to_replace = {
         "OVE_HOST": f"{ip}:8080",
-        "OPENVIDU_HOST": f"{ip}:4443",
-        "openvidu.publicurl": f"https://{ip}:4443",
+        # TODO: Point to the on-prem openvidu (keep as port 4443)
+        "OPENVIDU_HOST": "https://146.179.34.13:4443",
     }
 
     # Read the template file
@@ -53,9 +53,6 @@ def generate_docker_compose(template_file: str, ip: str, develop: bool = False) 
         docker_compose = yaml.safe_load(f)
 
     # Replace the values in the docker-compose.yml file
-    docker_compose["services"]["openvidu-openvidu-call"]["environment"][
-        "openvidu.publicurl"
-    ] = lines_to_replace["openvidu.publicurl"]
     docker_compose["services"]["ovehub-ove-apps"]["environment"][
         "OVE_HOST"
     ] = lines_to_replace["OVE_HOST"]
@@ -65,10 +62,6 @@ def generate_docker_compose(template_file: str, ip: str, develop: bool = False) 
     docker_compose["services"]["ovehub-ove-ui"]["environment"][
         "OVE_HOST"
     ] = lines_to_replace["OVE_HOST"]
-    # Also replace the openvidu-call version
-    docker_compose["services"]["openvidu-openvidu-call"][
-        "image"
-    ] = "openvidu/openvidu-call:2.28.0"
 
     # Add the dash app to to docker-compose.yml file
     logging.info("Adding dash app to docker-compose.yml...")

--- a/configure.py
+++ b/configure.py
@@ -72,7 +72,7 @@ def generate_docker_compose(template_file: str, ip: str, develop: bool = False) 
     # Also replace the openvidu-call version
     docker_compose["services"]["openvidu-openvidu-call"][
         "image"
-    ] = "openvidu/openvidu-call:2.12.0"
+    ] = "openvidu/openvidu-call:2.28.0"
 
     # Add the dash app to to docker-compose.yml file
     logging.info("Adding dash app to docker-compose.yml...")

--- a/docker-compose.setup.ove.yml
+++ b/docker-compose.setup.ove.yml
@@ -10,16 +10,16 @@ services:
     image: ovehub/ove-external-tuoris:v0.1
     restart: unless-stopped
     ports:
-    - "7080:7080"
+      - "7080:7080"
 
   openvidu-openvidu-call:
-    image: openvidu/openvidu-call:2.7.0
+    image: openvidu/openvidu-call:2.7.0 # Need to use openvidu/openvidu-call:2.28.0
     restart: unless-stopped
     ports:
-    - "4443:4443"
+      - "443:443"
     environment:
       openvidu.secret: "MY_SECRET"
-      openvidu.publicurl: "https://0.0.0.0:4443"
+      openvidu.publicurl: "0.0.0.0:443"
 
   ovehub-ove-service-persistence-inmemory:
     image: ovehub/ove-service-persistence-inmemory:0.2.1
@@ -35,7 +35,7 @@ services:
     environment:
       OVE_HOST: "0.0.0.0:8080"
       TUORIS_HOST: "0.0.0.0:7080"
-      OPENVIDU_HOST: "0.0.0.0:4443"
+      OPENVIDU_HOST: "https://0.0.0.0:443"
       OPENVIDU_SECRET: "MY_SECRET"
 
   ovehub-ove-ui:
@@ -56,6 +56,6 @@ services:
       - ovehub-ove-service-persistence-inmemory
       - ovehub-ove-ui
     ports:
-    - "8080:8080"
+      - "8080:8080"
     volumes:
       - ./config/default.conf:/etc/nginx/conf.d/default.conf:ro

--- a/docker-compose.setup.ove.yml
+++ b/docker-compose.setup.ove.yml
@@ -6,21 +6,11 @@ services:
     volumes:
       - ./config/Spaces.json:/usr/src/app/packages/ove-core/dist/client/Spaces.json:ro
 
-  openvidu-openvidu-call:
-    image: openvidu/openvidu-call:2.28.0
-    restart: unless-stopped
-    ports:
-      - "443:443"
-    environment:
-      openvidu.secret: "MY_SECRET"
-      openvidu.publicurl: "0.0.0.0:443"
-
   ovehub-ove-apps:
     image: ovehub/ove-apps:0.5.0
     restart: unless-stopped
     depends_on:
       - ovehub-ove-ove
-      - openvidu-openvidu-call
     environment:
       OVE_HOST: "0.0.0.0:8080"
       OPENVIDU_HOST: "https://0.0.0.0:443"

--- a/docker-compose.setup.ove.yml
+++ b/docker-compose.setup.ove.yml
@@ -6,14 +6,8 @@ services:
     volumes:
       - ./config/Spaces.json:/usr/src/app/packages/ove-core/dist/client/Spaces.json:ro
 
-  ovehub-ove-external-tuoris:
-    image: ovehub/ove-external-tuoris:v0.1
-    restart: unless-stopped
-    ports:
-      - "7080:7080"
-
   openvidu-openvidu-call:
-    image: openvidu/openvidu-call:2.7.0 # Need to use openvidu/openvidu-call:2.28.0
+    image: openvidu/openvidu-call:2.28.0
     restart: unless-stopped
     ports:
       - "443:443"
@@ -21,20 +15,14 @@ services:
       openvidu.secret: "MY_SECRET"
       openvidu.publicurl: "0.0.0.0:443"
 
-  ovehub-ove-service-persistence-inmemory:
-    image: ovehub/ove-service-persistence-inmemory:0.2.1
-    restart: unless-stopped
-
   ovehub-ove-apps:
     image: ovehub/ove-apps:0.5.0
     restart: unless-stopped
     depends_on:
       - ovehub-ove-ove
-      - ovehub-ove-external-tuoris
       - openvidu-openvidu-call
     environment:
       OVE_HOST: "0.0.0.0:8080"
-      TUORIS_HOST: "0.0.0.0:7080"
       OPENVIDU_HOST: "https://0.0.0.0:443"
       OPENVIDU_SECRET: "MY_SECRET"
 
@@ -53,7 +41,6 @@ services:
     depends_on:
       - ovehub-ove-ove
       - ovehub-ove-apps
-      - ovehub-ove-service-persistence-inmemory
       - ovehub-ove-ui
     ports:
       - "8080:8080"

